src/buffer/out/LineRendition.hpp

```diff
@@ -21,6 +21,13 @@ enum class LineRendition : uint8_t
     DoubleHeightBottom
 };
 
+constexpr til::rect ScreenToBufferLine(const til::rect& line, const LineRendition lineRendition)
+{
+    // Use shift right to quickly divide the Left and Right by 2 for double width lines.
+    const auto scale = lineRendition == LineRendition::SingleWidth ? 0 : 1;
+    return { line.left >> scale, line.top, line.right >> scale, line.bottom };
+}
+

```

Comment:
I think the right hand side may be off by one when you're dealing with exclusive coordinates with an odd value. For example, if the right screen coordinate is 7 exclusive (6 inclusive), that should map to a buffer coordinate of 4 exclusive (3 inclusive). A simple right shift works for inclusive coordinates (`6 >> 1 = 3`), but not for exclusive coordinates (`7 >> 1 = 3`, but we want 4).

Reply:
Hmm... I think the current approach is valid as well, in a certain way. Currently, these two related functions round down the width of the buffer:
* `TextBuffer::GetLineWidth`
* `ROW::GetReadableColumnCount`

If this function where to round up the `right` coordinate, then passing a viewport sized `til::rect` will end up having a different size than the size reported by the above two functions.

I do prefer your suggestion, but do we need to change other code first before we can round exclusive coordinates up here?

Reply:
Oh, I guess this means we've had this inconsistency for a while right? Because the `inclusive_rect` variant of `ScreenToBufferLine` may have reported an inclusive `.right` of 59 while the above two functions reported a max. width of 59. Hmm... I'm not sure how to best resolve this.

Reply:
I think they're expected to be inconsistent. `GetLineWidth` tells you how many buffer columns you can fit on the screen, so it has to round down if only half of the last column will fit. But `ScreenToBufferLine` is used to determine how much of the buffer is required to cover a given screen area. If you round down, you won't get enough buffer content, and the last screen cell won't be updated.

Reply:
Hmm, yeah I see what you mean. However, I'm also a little squirmish about what you said. I don't want this function end up being the reason why there's a off-by-1 error at some point in the future (for instance if we were to use the resulting `rect` to slice into a attribute vector which has 1 item too few). I think this function is unused so far and so I'll remove it for now.

Idle thoughts: Personally speaking, I would prefer if we could consistently use `til::rect` with its exclusive coordinates everywhere, even in areas of our code where inclusive coordinates would be a better fit, just so that everything works consistently. I wonder if this issue would go away with such a change as well... Probably not really, but it may still avoid some potential for confusion.